### PR TITLE
[Merged by Bors] - Revert fork choice if disk write fails

### LINKED ISSUE #2028: Failed to build beacon chain: Head block not found in store
## Description

Beacon chain was working properly on Pyrmont for a few days and then it stopped and doesn't start up again. 

## Version

Lighthouse/v1.0.3-65dcdc36+

## Present Behaviour

Beacon chain stops right after starting up with error: Failed to build beacon chain: Head block not found in store

```
Dec 01 08:35:21.582 INFO Lighthouse started                      version: Lighthouse/v1.0.3-65dcdc36+
Dec 01 08:35:21.582 INFO Configured for network                  name: pyrmont
Dec 01 08:35:21.582 INFO Data directory initialised              datadir: /var/lib/lighthouse
Dec 01 08:35:21.582 WARN Running HTTP server on port 5052
Dec 01 08:35:21.592 INFO Deposit contract                        address: 0x8c5fecdc472e27bc447696f431e425d02dd46a8c, deploy_block: 3743587
Dec 01 08:35:21.623 INFO Hot-Cold DB initialized                 split_state: 0xaebea146f4d6c06ad2240530650c14556cbc45f41eec4d245db4a186da111a22, split_slot: 90880, version: 2, service: freezer_db
Dec 01 08:35:21.624 INFO Starting beacon chain                   method: resume, service: beacon
Dec 01 08:35:31.905 INFO Block production enabled                method: json rpc via http, endpoints: ["http://127.0.0.1:8545"]
Dec 01 08:35:32.134 CRIT Failed to start beacon node             reason: Failed to build beacon chain: Head block not found in store
Dec 01 08:35:32.134 INFO Internal shutdown received              reason: Failed to start beacon node
Dec 01 08:35:32.134 INFO Shutting down..
```

Logs from before restarting it
```
Dec 01 07:43:47.258 INFO Sync state updated                      new_state: Syncing Finalized Chain, old_state: Searching syncing peers, service: sync
Dec 01 07:43:48.123 WARN BlockProcessingFailure                  outcome: MissingBeaconBlock(0xb866258c3d9956f6901712909729338cc2a1651383e12b4cdf9b202b328b0856), msg: unexpected condition in processing block.
Dec 01 07:43:49.001 INFO Syncing                                 est_time: --, distance: 1343 slots (4 hrs 28 mins), peers: 55, service: slot_notifier
Dec 01 07:44:01.000 INFO Syncing                                 est_time: --, distance: 1344 slots (4 hrs 28 mins), peers: 55, service: slot_notifier
Dec 01 07:44:13.000 INFO Syncing                                 est_time: --, distance: 1345 slots (4 hrs 29 mins), peers: 55, service: slot_notifier
Dec 01 07:44:14.550 WARN Batch failed to download. Dropping chain scoring peers, batch_epoch: 2840, score_adjustment: Low Tolerance Error, chain: 18141939244923768013, service: sync
Dec 01 07:44:15.641 WARN BlockProcessingFailure                  outcome: MissingBeaconBlock(0xb866258c3d9956f6901712909729338cc2a1651383e12b4cdf9b202b328b0856), msg: unexpected condition in processing block.
Dec 01 07:44:25.000 INFO Syncing                                 est_time: --, distance: 1346 slots (4 hrs 29 mins), peers: 55, service: slot_notifier
Dec 01 07:44:31.511 WARN Batch failed to download. Dropping chain scoring peers, batch_epoch: 2840, score_adjustment: Low Tolerance Error, chain: 16675040628877297081, service: sync
Dec 01 07:44:31.514 INFO Sync state updated                      new_state: Searching syncing peers, old_state: Syncing Finalized Chain, service: sync
Dec 01 07:44:31.565 INFO Sync state updated                      new_state: Syncing Finalized Chain, old_state: Searching syncing peers, service: sync
Dec 01 07:44:33.732 WARN BlockProcessingFailure                  outcome: MissingBeaconBlock(0xb866258c3d9956f6901712909729338cc2a1651383e12b4cdf9b202b328b0856), msg: unexpected condition in processing block.
Dec 01 07:44:37.000 INFO Syncing                                 est_time: --, distance: 1347 slots (4 hrs 29 mins), peers: 55, service: slot_notifier
Dec 01 07:44:42.479 WARN Batch failed to download. Dropping chain scoring peers, batch_epoch: 2840, score_adjustment: Low Tolerance Error, chain: 16675040628877297081, service: sync
Dec 01 07:44:42.482 INFO Sync state updated                      new_state: Searching syncing peers, old_state: Syncing Finalized Chain, service: sync
Dec 01 07:44:42.511 INFO Sync state updated                      new_state: Syncing Finalized Chain, old_state: Searching syncing peers, service: sync
Dec 01 07:44:43.134 WARN BlockProcessingFailure                  outcome: MissingBeaconBlock(0xb866258c3d9956f6901712909729338cc2a1651383e12b4cdf9b202b328b0856), msg: unexpected condition in processing block.
Dec 01 07:44:44.110 INFO Shutting down..
Dec 01 07:44:44.112 INFO Saved DHT state                         service: network
Dec 01 07:44:44.112 INFO Network service shutdown                service: network
Dec 01 07:44:44.459 INFO Saved beacon chain to disk              service: beacon
```

## Expected Behaviour

It should start up

### PR DESCRIPTION
**Pull request by michaelsproul**
## Issue Addressed

Closes #2028
Replaces #2059

## Proposed Changes

If writing to the database fails while importing a block, revert fork choice to the last version stored on disk. This prevents fork choice from being ahead of the blocks on disk. Having fork choice ahead is particularly bad if it is later successfully written to disk, because it renders the database corrupt (see #2028).

## Additional Info

* This mitigation might fail if the head+fork choice haven't been persisted yet, which can only happen at first startup (see #2067)
* This relies on it being OK for the head tracker to be ahead of fork choice. I figure this is tolerable because blocks only get added to the head tracker after successfully being written on disk _and_ to fork choice, so even if fork choice reverts a little bit, when the pruning algorithm runs, those blocks will still be on disk and OK to prune. The pruning algorithm also doesn't rely on heads being unique, technically it's OK for multiple blocks from the same linear chain segment to be present in the head tracker. This begs the question of #1785 (i.e. things would be simpler with the head tracker out of the way). Alternatively, this PR could just revert the head tracker as well (I'll look into this tomorrow).

### CODE REVIEW

**Comment by michaelsproul** [comment]
I've been thinking more about whether to reset the head tracker as well, and I think I'm leaning towards not resetting it. If we leave it intact then we won't forget about any blocks written to disk, and can prune them later. Compare this to if we do reset: where blocks get forgotten about on disk and potentially bloat the database indefinitely.

File: beacon_node/beacon_chain/src/beacon_chain.rs [hunk 1]
```diff
@@ -1715,13 +1739,37 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
...
+            );
+            match Self::load_fork_choice(self.store.clone())? {
+                Some(persisted_fork_choice) => {
+                    *fork_choice = persisted_fork_choice;
+                }
+                None => {
+                    crit!(
+                        self.log,
+                        "No stored fork choice found to restore from";
+                        "warning" => "The database is likely corrupt now"
```
Comment (paulhauner) [inline beacon_node/beacon_chain/src/beacon_chain.rs:1767]: Perhaps we should mention `--purge-db` here?

**Review by paulhauner — APPROVED** [review]
This seems like an easy and effective solution, nice idea!

Apart from the head tracker, I don't think this creates any new assumptions. It will still present as the same error to downstream components (p2p, api ,etc), which is good. That being said, reverting fork choice will have *some* effects on downstream components, e.g: 

- Peers to be banned on the p2p network because we may interpret some blocks as "parent unknown" and therefore not a valid chain. I think this is a little contrived though.
- Attestations from p2p and the api might be rejected. This could result in missed attestations by a validator.

Although these outcomes aren't desirable, they come about as the result of an IO error that should only happen when there's underlying hardware issues. These changes allow us to go from stalling to staying online, which is a huge leap. Going from staying online to operating perfectly would be a lot of work with marginal gains, IMO.

I'm happy to merge this. I made one comment, it's not a blocker though.

File: beacon_node/beacon_chain/src/beacon_chain.rs (hunk 1 above)
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: Done!

**Pull request closed without merging**